programs/ido-pool/src/lib.rs

```diff
@@ -183,19 +227,24 @@ pub mod ido_pool {
         }
 
         // Calculate watermelon tokens due.
-        let watermelon_amount = (amount as u128)
-            .checked_mul(ctx.accounts.pool_watermelon.amount as u128)
-            .unwrap()
-            .checked_div(ctx.accounts.redeemable_mint.supply as u128)
-            .unwrap();
+        let watermelon_amount = cmp::min(
+            ctx.accounts.pool_watermelon.amount,
+            (amount as u128)
+                .checked_mul(ctx.accounts.pool_watermelon.amount as u128)
+                .unwrap()
+                .checked_div(ctx.accounts.redeemable_mint.supply as u128)
+                .unwrap()
+                .try_into()
+                .unwrap(),
+        );
```

Comment:
i dont think the `cmp::min()` is necessary since int division is floor division and `amount <= redeemable_mint.supply` so `watermelon_amount <= pool_watermelon.amount`. Not 100% sure about this.

However seeing this check did make me realize another possible issue. If the price of watermelon ever goes above 1 USDC, it's possible for `watermelon_amount` to be 0 since `pool_watermelon.amount < redeemable_mint.supply`. Currently if this happens, this unlucky user will just get his redeemables burnt but get 0 watermelon tokens in return. We might not need to worry about this case at this point since it seems unlikely that the price will pump to 1 USDC but we should keep this in mind for now, maybe set a max_token price that's <= 1 USDC

Reply:
- On my previous auction contract, one of the issues happened
The remaining pool_watermelon.amount was 180000000000 (ether_value 18.0000000000) but the calculated amount was 180000000001
So the redeem action was failed because of this issue: the transfer amount is bigger than the balance
So I have used the `cmp::min()` to avoid this kind of issue

Reply:
- The above calculation is working with the wei value so even the price of the watermelon is above the 1 USDC, the redeemable amount wouldn't be zero

(amount: 2_000) * (pool_watermelon: 300_000_000) / (redeemable_mint: 600_000_000) = (user_watermelon: 1_000)
In this case, the Watermelon price is 2 USDC
The user has purchased only 0.000002 USDC (2000 in wei) and redeemable amount is 0.000001 Watermelon (1000 in wei)

Reply:
> * On my previous auction contract, one of the issues happened
>       The remaining pool_watermelon.amount was 180000000000 (ether_value 18.0000000000) but the calculated amount was 180000000001
>       So the redeem action was failed because of this issue: the transfer amount is bigger than the balance
>       So I have used the `cmp::min()` to avoid this kind of issue

oh wow was it with this program? Or was this another auction program?

Reply:
> * The above calculation is working with the wei value so even the price of the watermelon is above the 1 USDC, the redeemable amount wouldn't be zero
> 
> 
> (amount: 2_000) * (pool_watermelon: 300_000_000) / (redeemable_mint: 600_000_000) = (user_watermelon: 1_000) In this case, the Watermelon price is 2 USDC The user has purchased only 0.000002 USDC (2000 in wei) and redeemable amount is 0.000001 Watermelon (1000 in wei)

what if amount = 1_999 in this case?

Reply:
> > * On my previous auction contract, one of the issues happened
> >   The remaining pool_watermelon.amount was 180000000000 (ether_value 18.0000000000) but the calculated amount was 180000000001
> >   So the redeem action was failed because of this issue: the transfer amount is bigger than the balance
> >   So I have used the `cmp::min()` to avoid this kind of issue
> 
> oh wow was it with this program? Or was this another auction program?

It was on another program (the auction contract on Ethereum)

Reply:
> > * The above calculation is working with the wei value so even the price of the watermelon is above the 1 USDC, the redeemable amount wouldn't be zero
> > 
> > (amount: 2_000) * (pool_watermelon: 300_000_000) / (redeemable_mint: 600_000_000) = (user_watermelon: 1_000) In this case, the Watermelon price is 2 USDC The user has purchased only 0.000002 USDC (2000 in wei) and redeemable amount is 0.000001 Watermelon (1000 in wei)
> 
> what if amount = 1_999 in this case?

It would be 999

Reply:
> > > * The above calculation is working with the wei value so even the price of the watermelon is above the 1 USDC, the redeemable amount wouldn't be zero
> > > 
> > > (amount: 2_000) * (pool_watermelon: 300_000_000) / (redeemable_mint: 600_000_000) = (user_watermelon: 1_000) In this case, the Watermelon price is 2 USDC The user has purchased only 0.000002 USDC (2000 in wei) and redeemable amount is 0.000001 Watermelon (1000 in wei)
> > 
> > 
> > what if amount = 1_999 in this case?
> 
> It would be 999

oh sorry lol i meant `amount = 1`

Reply:
zero :)